028-service-binding.md

```diff
@@ -0,0 +1,379 @@
+    ssl.truststore.crt: #  Strimzi cluster CA certificate (copied from ca.crt in <clustername>-cluster-ca-cert secret)
+    ssl.truststore.p12: #  PKCS #12 archive file for Strimzi cluster CA certificate (copied from ca.p12 in <clustername>-cluster-ca-cert secret)
+    ssl.truststore.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file (copied from ca.password in <clustername>-cluster-ca-cert secret)
+    # Provided if selected user is SCRAM auth:
+    username: # SCRAM username
+    password: # SCRAM password
+    sasl.jaas.config: # sasl jaas config string for use by Java applications
+    sasl.mechanism: SCRAM-SHA-512
+    # Provided if selected user is mTLS:
+    ssl.keystore.p12: # client certificate for the consuming client PKCS #12 archive file for storing certificates and keys (copied from user.p12 in KafkaUser secret)
+    ssl.keystore.password: # password for protecting the client certificate PKCS #12 archive file (copied from user.password in KafkaUser secret)
+    ssl.keystore.crt: # certificate for the consuming client signed by the clients' CA (copied from user.crt in KafkaUser secret)
+    ssl.keystore.key: # private key for the consuming client (copied from user.key in KafkaUser secret)
+```
+
+Naming of secret keys:
+
+This proposal previously considered choosing secret keys that could be reused by other (non-Kafka) operators that use mutualTLS or SASL SCRAM authentication. However, given 
+that Kafka applications generally have a well known list of configuration options they provide, the proposal instead uses those well-known keys. This will make it clearer 
+to application developers what the different keys should be used for.
+
+### Implementation
+
+This proposal will result in a new `KafkaConnectionAssemblyOperator` class being added to the `cluster-operator` directory in the `strimzi-kafka-operator` repo. The class will be 
+deployed as a separate Verticle and have its own reconcile loop.
+
+The actions taken during the reconcile loop will be:
+1. The operator can be deployed at cluster-scope where it will watch resources across all namespaces, or in a single namespace, where it will only watch resources in that namespace.
+2. On start-up the operator will watch all resources of Kind `KafkaConnection`.
+3. For any `KafkaConnection` resource present it will add a status of NotReady to that CR.
+4. For each `KafkaConnection` resource it will perform a GET request on the Kafka cluster that is 
+referenced in the `KafkaConnection` and pick out the configured listeners.
+   1. If this or any subsequent GET requests fail due to a 401 (unauthorised) error, the status reason will be updated to reflect this.
+5. If there are multiple listeners listed in the `Kafka` CR with the type `internal`, the operator 
+will filter the list by comparing the `tls` and `authentication` properties in the `Kafka` and `KafkaUser` CRs.
+6. If there are still multiple listeners it will pick one at random.
```

Comment:
This non-determinism bothers me because it means users could experience weird behaviour. E.g. if you had two `KafkaConnection` CRs that differed only in the `KafkaUser` the pointed to then the resulting bindings might end up selecting different listeners, which would be inconsistent from the users PoV. I think we need a way to make it deterministic. The best way imho would be to require selection by listener name if there is ambiguity. Alternatives include using the first listener in the filtered list, or sorting the filtered list by name and picking the first, but imho they're inferior because they're too implicit and prone to breakage by seemingly innocuous changes in the `Kafka` CR.

Reply:
I hadn't thought of the inconsistency between two different `KafkaConnection` CRs. Could we not make listener selection a requirement but add a warning to the status of the `KafkaConnection` CR if the operator had to make a choice between multiple options?

Reply:
It has to be deterministic not just to not differ between two resources. If it is random, it might change in every reconciliation and just loop in a circle. I think we should just pick a sorting rule ... lowest port number, listener names by alphabet etc.

Reply:
The mechanism described in step 7 will stop it cycling, the intention was only to pick a random one if there isn't one listed in the status already. That way even if the port number changes later the listener chosen will stay as before.

Reply:
How does that deal with the changes in the Kafka CR? I don't know, random seems weird -> I think it is easy to come up with some fixed logic which works in most cases.

Reply:
Ok, I think sorting based on listener name is the simplest and least likely to change. I will update the proposal to use that